### PR TITLE
Using bytes orderData instead of array of order inputs

### PR DESCRIPTION
**Pull request by josojo**
This PR is making sure that the orders are no longer emitted with separate variables buyToken, sellToken, buyAmounts, sellAmount
but instead, are just emitted as one order blob
```
event SellOrder(
        uint auctionId,
        uint16 slotIndex,
        uint16 accountId,
        uint8 buyToken,
        uint8 sellToken,
        uint96 buyAmount,
        uint96 sellAmount
    );
to
event SellOrder(
        uint auctionId,
        uint16 slotStartIndex,
        uint16 accountId,
        bytes orderData
    );
```
it also makes sure that placing order is done with packedOrderdata as well.

testplan:
unit tests only

### CODE REVIEW

File: contracts/SnappAuction.sol [hunk 1]
```diff
@@ -170,7 +175,10 @@ contract SnappAuction is SnappBase {
         );
         auctions[auctionIndex].shaHash = nextAuctionHash;
 
-        emit SellOrder(auctionIndex, auctions[auctionIndex].size, accountId, buyToken, sellToken, buyAmount, sellAmount);
+        emit SellOrder(
+            auctionIndex, auctions[auctionIndex].size, accountId,
+            abi.encode((uint(buyToken) << 16) + (uint(sellToken) << 24) + (uint(sellAmount) << 32) + (uint(buyAmount) << 128))
```
Comment (fleupold) [inline contracts/SnappAuction.sol]: I don't think this event should change, but if let's please reuse the `encodeOrder` function.

File: contracts/SnappAuction.sol [hunk 2]
```diff
@@ -126,12 +117,26 @@ contract SnappAuction is SnappBase {
         ) {
             createNewPendingBatch();
         }
-
+        bytes32 orderHash;
+        bytes memory orderData;
         for (uint i = 0; i < numOrders; i++) {
+            orderData = packedOrders.slice(26*i, 26);
```
Comment (fleupold) [inline contracts/SnappAuction.sol]: Please extract the other code that does this into a helper function and reuse instead of rewriting it.

**Review by fleupold — CHANGES_REQUESTED** [review]
Let's reuse some of the solidity code we have written earlier here. Also, I thought that a single order can stay in its old form? What are your thoughts?

File: contracts/SnappAuction.sol (hunk 1 above)
Comment (josojo) [reply contracts/SnappAuction.sol]: The code will be killed anyways(including the event) with your PR, right?

Comment (josojo) [reply contracts/SnappAuction.sol]: I will rebase it on your PR.

File: contracts/SnappAuction.sol (hunk 2 above)
Comment (josojo) [reply contracts/SnappAuction.sol]: Good suggestion

File: contracts/SnappAuction.sol [hunk 3]
```diff
@@ -32,19 +32,13 @@ contract SnappAuction is SnappBase {
         uint auctionId,
         uint16 slotIndex,
         uint16 accountId,
-        uint8 buyToken,
-        uint8 sellToken,
-        uint96 buyAmount,
-        uint96 sellAmount
+        bytes packedOrders
```
Comment (bh2smith) [inline contracts/SnappAuction.sol]: This should be singular `packedOrder` or even `orderData`

File: test/snapp_utils.js [hunk 4]
```diff
@@ -53,12 +53,12 @@ const encodePacked_16_8_128 = function(a, b, c) {
...
-  return Buffer.from(uint8(buyToken) + uint8(sellToken) +  uint96(buyAmount) + uint96(sellAmount), "hex")
+  return Buffer.from(uint96(buyAmount) + uint96(sellAmount) + uint8(sellToken) + uint8(buyToken) , "hex")
 }
 
 module.exports = {
   falseArray,
   isActive,
   encodePacked_16_8_128,
-  encodeOrder
+  encodeOrder
```
Comment (bh2smith) [inline test/snapp_utils.js]:
```suggestion
  encodeOrder
```

Comment (bh2smith) [reply test/snapp_utils.js]: This would be why the build is failing! "Trailing whitespace"

File: contracts/SnappAuction.sol (hunk 3 above)
Comment (bh2smith) [reply contracts/SnappAuction.sol]:
```suggestion
        bytes packedOrder
```

File: contracts/SnappAuction.sol [hunk 5]
```diff
@@ -149,30 +136,18 @@ contract SnappAuction is SnappBase {
...
-                encodeOrder(accountId, buyToken, sellToken, buyAmount, sellAmount)
-            )
-        );
-        auctions[auctionIndex].shaHash = nextAuctionHash;
-
-        emit SellOrder(auctionIndex, auctions[auctionIndex].size, accountId, buyToken, sellToken, buyAmount, sellAmount);
-        // Only increment size after event (so it is emitted as an index)
-        auctions[auctionIndex].size++;
+    ) public {
+        // Ignore first 4 bytes padding and last two bytes accountId
```
Comment (bh2smith) [inline contracts/SnappAuction.sol]: Why are first 4 bytes being ignored?

File: contracts/SnappAuction.sol [hunk 6]
```diff
@@ -120,25 +120,7 @@ contract SnappAuction is SnappBase {
...
-            assembly {  // solhint-disable no-inline-assembly
-                sellAmount := mload(add(add(orderData, 0xc), 14))
-            }
-            orderHash = sha256(
-                abi.encodePacked(
-                    orderHash,
-                    encodeOrder(accountId, buyToken, sellToken, buyAmount, sellAmount)
-                )
-            );
+            orderhash = calculateNextOrderHashIteration(packedOrders.slice(26*i, 26), orderhash);
```
Comment (bh2smith) [inline contracts/SnappAuction.sol]: What happened with `bytes memory orderData` from above? this looks like an unused variable now?
Do we know this works if the auction index is increased in the middle of this loop?

Also, I think this function name is way too long! Suggest something like `parseOrder`

File: contracts/SnappAuction.sol [hunk 7]
```diff
@@ -262,6 +225,27 @@ contract SnappAuction is SnappBase {
         // solhint-disable-next-line max-line-length
         return bytes32(uint(accountId) + (uint(buyToken) << 16) + (uint(sellToken) << 24) + (uint(sellAmount) << 32) + (uint(buyAmount) << 128));
     }
+    function calculateNextOrderHashIteration(bytes orderData, bytes32 previousHash) internal
+        returns(bytes32)
+    {
+        buyToken = BytesLib.toUint8(orderData, 0);
+        sellToken = BytesLib.toUint8(orderData, 1);
+
+        buyAmount;
```
Comment (bh2smith) [inline contracts/SnappAuction.sol]: Do we not need to declare the type here?

File: contracts/SnappAuction.sol [hunk 8]
```diff
@@ -262,6 +225,27 @@ contract SnappAuction is SnappBase {
...
+        assembly {  // solhint-disable no-inline-assembly
+            buyAmount := mload(add(add(orderData, 0xc), 2))
+        }
+        sellAmount;
+        assembly {  // solhint-disable no-inline-assembly
+            sellAmount := mload(add(add(orderData, 0xc), 14))
+        }
+        return nextHash = sha256(
```
Comment (bh2smith) [inline contracts/SnappAuction.sol]: Do we need to declare nextHash if it is being returned anyway?

**Review by bh2smith — APPROVED** [review]
I think there are a few things that need changing and maybe some justification, but it it starting to look very nice.

**Review by bh2smith — CHANGES_REQUESTED** [review]
I think there are a few things that need changing and maybe some justification, but it it starting to look very nice.

File: contracts/SnappAuction.sol (hunk 6 above)
Comment (bh2smith) [reply contracts/SnappAuction.sol]: Suggest `updateOrderHash`

File: contracts/SnappAuction.sol [hunk 9]
```diff
@@ -268,6 +227,28 @@ contract SnappAuction is SnappBase {
         return bytes32(uint(accountId) + (uint(buyToken) << 16) + (uint(sellToken) << 24) + (uint(sellAmount) << 32) + (uint(buyAmount) << 128));
     }
 
+    function calculateNextOrderHashIteration(bytes orderData, bytes32 previousHash) internal
```
Comment (bh2smith) [inline contracts/SnappAuction.sol]:
```suggestion
    function calculateNextOrderHashIteration(bytes memory orderData, bytes32 previousHash) internal
```

File: contracts/SnappAuction.sol [hunk 10]
```diff
@@ -30,21 +30,15 @@ contract SnappAuction is SnappBase {
 
     event SellOrder(
         uint auctionId,
-        uint16 slotIndex,
+        uint16 slotStartIndex,
```
Comment (fleupold) [inline contracts/SnappAuction.sol]: why this change?

File: contracts/SnappAuction.sol [hunk 11]
```diff
@@ -168,38 +154,23 @@ contract SnappAuction is SnappBase {
...
-            emit SellOrder(
-                auctionIndex, auctions[auctionIndex].size, accountId, buyToken, sellToken, buyAmount, sellAmount
-            );
+
             // Only increment size after event (so it is emitted as an index)
             auctions[auctionIndex].size++;
         }
+        emit SellOrder(
+            auctionIndex, auctions[auctionIndex].size, accountId, packedOrders
+            );
```
Comment (fleupold) [inline contracts/SnappAuction.sol:173]: nit: wrong indent

File: contracts/SnappAuction.sol [hunk 12]
```diff
@@ -253,6 +224,29 @@ contract SnappAuction is SnappBase {
...
+
+        uint8 sellToken = BytesLib.toUint8(orderData, 24);
+        uint8 buyToken = BytesLib.toUint8(orderData, 25);
+
+        return sha256(
+            abi.encodePacked(
+                    previousHash,  // TODO - Below todo will affect this.
+                    encodeOrder(accountId, buyToken, sellToken, buyAmount, sellAmount)
+                )
+            );
```
Comment (fleupold) [inline contracts/SnappAuction.sol]: ditto

File: contracts/SnappAuction.sol [hunk 13]
```diff
@@ -253,6 +224,29 @@ contract SnappAuction is SnappBase {
...
+        assembly {  // solhint-disable no-inline-assembly
+            sellAmount := mload(add(add(orderData, 0xc), 12))
+        }
+
+        uint8 sellToken = BytesLib.toUint8(orderData, 24);
+        uint8 buyToken = BytesLib.toUint8(orderData, 25);
+
+        return sha256(
+            abi.encodePacked(
+                    previousHash,  // TODO - Below todo will affect this.
```
Comment (fleupold) [inline contracts/SnappAuction.sol]: below todo dosn't make sense anymore

File: contracts/SnappAuction.sol [hunk 14]
```diff
@@ -253,6 +224,29 @@ contract SnappAuction is SnappBase {
         return bytes32(uint(accountId) + (uint(buyToken) << 16) + (uint(sellToken) << 24) + (uint(sellAmount) << 32) + (uint(buyAmount) << 128));
     }
 
+    function calculateNextOrderHashIteration(bytes memory orderData, bytes32 previousHash, uint16 accountId) internal view
```
Comment (fleupold) [inline contracts/SnappAuction.sol:227]: nit, naming suggestion: orderHashFromPreviousWithNew[Order]

File: contracts/SnappAuction.sol [hunk 15]
```diff
@@ -30,21 +30,15 @@ contract SnappAuction is SnappBase {
...
     event SellOrder(
         uint auctionId,
-        uint16 slotIndex,
+        uint16 slotStartIndex,
         uint16 accountId,
-        uint8 buyToken,
-        uint8 sellToken,
-        uint96 buyAmount,
-        uint96 sellAmount
+        bytes packedOrder
```
Comment (fleupold) [inline contracts/SnappAuction.sol:35]: Not sure what happened to my comment from earlier, but do we really want this change? this will require a bunch more changes in the listener, and if at all necessary it should maybe happen in a separate PR?

File: contracts/SnappAuction.sol (hunk 13 above)
Comment (josojo) [reply contracts/SnappAuction.sol]:
```suggestion
                    previousHash,
```

File: contracts/SnappAuction.sol (hunk 15 above)
Comment (bh2smith) [reply contracts/SnappAuction.sol:35]: I don't understand the problem here! Is it the plural vs non-plural?

Comment (josojo) [reply contracts/SnappAuction.sol:35]: @fleupold This whole Pr was about changing the sellorderEvent, wasn't it? Do you think having a seperate Event would be better?

**Review by bh2smith — APPROVED** [review]
Looks good to me!

File: contracts/SnappAuction.sol (hunk 15 above)
Comment (fleupold) [reply contracts/SnappAuction.sol:35]: I was under the impression the whole PR was about changing the **StandingOrderEvent**. The sellOrder event doesn't necessarily need to be adjusted.

**Comment by fleupold** [comment]
If we also adjust the normal, unreserved sell order event, we will have to change the listener and the driver, before https://github.com/gnosis/dex-services/pull/132 is unblocked.

**Comment by josojo** [comment]
Okay, then there was a misunderstanding. Sry. :D it was not really unblocking you, as this Pr was much bigger in scope as intended... Damn, next time I have to read it better. Now I will split this PR into two: One without changing the sellorder event

**Comment by fleupold** [comment]
Outdated
